text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,297 @@
+We still have the restriction that new features can be added below, but this restriction is no longer a problem:
+there is no point of adding such a new minimal feature because there is nothing left to `cfg`-out!
+
+This solution is more mathematically elegant, but it seems harder to implement.
+It is unclear how Cargo could require the Rust code to obey this property without new infra like the portability lint.
+
+Finally, in a sense these variations are actually isomorphic.
+If no one can request the empty feature set, no one can observe what library interface that corresponds to.
+There is no reason to conclude that it *isn't* the empty library.
+
+## RFC [#3146](https://github.com/rust-lang/rfcs/pull/3146)
+
+This more more expensive solution aimed to solve the "running out of good names" problem this has.
+
+Returning to our commutative diagram, instead of
+
+- removing the ⊥ feature downset (this proposal)
+
+- mandating the ⊥ feature downset correspond to ⊥ the empty library interface (first alternative, extra rules on the vertical mappings in diagrams)
+
+We can instead
+
+- allow customizing the mapping from old feature names to new feature names (customizing top horizontal mapping)
+
+This works and allows giving things the "right" names" every time, but is much work work to implement.
+
+One [point brought up in the comments](https://github.com/rust-lang/rfcs/pull/3146#issuecomment-1273473456) is that punning on the lower version bound to get the feature namespace is a bit tricky.
+In particular, it makes manually adjusting versions bounds risky because without consulting package definitions it is unclear how the feature list will be interpretted.
+I  think that is compelling reason not to do that, but instead version the feature namespace separately.
+
+Overal, I think it might make sense to do some sort of "name indirection" along these lines (but addressing the above problem), but it is better to start with this RFC before getting to that.
+My thinking about this is basically "walk before run".
+Today, that would be lots of work for an obscure problem, and a solution that downstream crates need to know about because it affects how dependencies are written.
+However if we do this first, "opt-out features" are poised to become much more popular now that they are no longer a compatibility hazard.
+We can evaluate both that new popularity, and the incidence of the "bad name" problem in *practice*, and use that to justify or not justify a more heavyweight solutions.
+
+## RFC [#3283](https://github.com/rust-lang/rfcs/pull/3283)
+
+This proposal adds a new notion of a "feature base", which aims to essential provide more `default-features = false` options rather than just one forever.
+`default-features = false` is ret-conned as the first feature base.
+
+My view is that that RFC is pretty close to being this one in disguise.
```

Comment:
To me this RFC seems like just a special case of RFC 3283. Specifically, the case where the set of base features includes all available features. And because it is just one special case, it has problems that 3283 didn't have, notably the problems described in "bad names over time". 

Personally, I'd rather see a more general solution where you specify multiple sets of features where at least one must be included, so you could for example say you need at least one of "http1" and "http2"  and at least one of a set of features for TLS support (rustls, native-tls, etc.).

Reply:
Hmm I don't think this is special case to that. Consider if `default` and `default-features = <bool>` and all that didn't exist. This would still be solution.

It is probably best to imagine 3 parts:

1. This to fix the original problem.
2. Some deprecation cycle with editions to get rid of `default-features = <bool>`
3. Some namespacing / indirection so we can use good names forever.

Each builds on the last, but the former steps stand alone.

Remember Rust itself has the same name issue! If you end up making a new version function which deserves the name more than the old version and don't want to make a breaking change...to bad. If it's not a show-stopper there, why should it be here?

> so you could for example say you need at least one

Disjunctive ("or", not "and") dependencies are completely unrelated to what is being discussed.

Reply:
If we try to solve the problem *and* deal with various back compact issues at the same (not library, but Cargo itself back compat), I think we'll just melt our brains.

*First* why should agree on an ideal design*, *second* we should try to figure out a migration plan to get there, *third* we should adjust the destination of it's not possible.

Reply:
> Hmm I don't think this is special case to that. Consider if default and default-features = <bool> and all that didn't exist. This would still be solution.

Unless I am missing some subtle distinction, having `empty-features = false` with this RFC is effectively equivalent to RFC 3283, where every feature in the crate is included in the `feature-bases`. Sure, the syntax is a little different because in 3283 you would need to put one of the features in the `base` attribute rather than `features` attribute, but that is a very surface-level detail. In both cases, you have to specify at least one feature. However, RFC 3283 is strictly more flexible, because rather than always requiring you to require one of _any_ feature in the crate. You can limit the requirement to a subset of the features. I don't see any argument in this RFC on why having a boolean to require "at least one feature" is a superior option to having a set of features, and requiring at least feature from that set (or possibly multiple such sets). 

> Remember Rust itself has the same name issue! If you end up making a new version function which deserves the name more than the old version and don't want to make a breaking change...to [sic] bad.

It's not quite the same. It isn't just that you can't use the name of an existing feature for a new feature. It is that if you add a new feature that was previously part of the "base", then it can potentially double the number of features that you need. If you do this multiple times, it can create quite mess, and you'll end with things like `foo-without-bar-without-baz-without-soup`.